sysinstall/sysmerge.8
.Dd $Mdocdate: February 14 2016 $
.Dt SYSMERGE 8
.Os
.Sh NAME
.Nm sysmerge
.Nd upgrade current operating system from a sysroot
.Sh SYNOPSIS
.Nm sysmerge
.Op Fl w
.Op Fl \-booting
.Op Fl \-cancel
.Op Fl \-wait
.Ar source
.Sh DESCRIPTION
.Nm
upgrades the current operating system by copying the system files from the
specified
.Ar source
directory (usually a sysroot) onto the current root filesystem.  This is meant
to be used when building the operating system from source as described in
.Xr development 7 .
.Pp
.Nm
installs the
.Sy system
manifest from the tix repository in the
.Ar source
directory, as well as all the ports found.  The
.Xr initrd 7
is regenerated using
.Xr update-initrd 8 .
If grub is enabled in
.Xr upgrade.conf 5 ,
then the bootloader is reinstalled and reconfigured as needed.
.Pp
.Nm
is an automatic and non-interactive upgrade.  It is meant to be used as part of
the development process to upgrade to locally built versions.  It does not
contain compatibility to help upgrade across larger changes, or to easily handle
changes to configuration file formats, or other caveats.  It places a burden on
the system administrator/developer to know the system changes in detail and to
do any necessary tasks manually.
.Pp
The
.Xr sysupgrade 8
program is by contrast an interactive program, meant to help upgrading across
much larger development distances.  It contains compatibility to automatically
handle changes in configuration files and other larger changes.
.Pp
The options are as follows:
.Bl -tag -width "12345678"
.It Fl \-booting
It's boot time, complete the system upgrade that was delayed.  This is meant to
be used by
.Xr init 8
through the
.Sy chain-merge
boot target.  This installs the
.Pa /sysmerge
directory onto the root filesystem and removes the
.Pa /sysmerge
directory.
.It Fl \-cancel
Cancel a pending upgrade that would trigger on the next boot.  Remove the
.Pa /sysmerge
directory and restore the old
.Xr kernel 7
and
.Xr initrd 7 .
.It Fl w , Fl \-wait
Wait until the next boot to complete the upgrade, rather than finishing it now.
This installs into the
.Pa /sysmerge
directory instead and replaces the
.Xr kernel 7
with the new kernel
and
.Xr initrd 7
with an initrd that runs
.Sy /sysmerge/sbin/sysmerge --booting
on boot through the
.Sy chain-merge
.Xr init 8
boot target.  Backups are made of the
.Xr kernel 7
and
.Xr initrd 7
such that the operation can be rolled back.
.El
.Sh FILES
.Bl -tag -width "/boot/sortix.initrd.sysmerge.orig" -compact
.It Pa /boot/sortix.bin
system
.Xr kernel 7
.It Pa /boot/sortix.bin.sysmerge.orig
system
.Xr kernel 7
(backup)
.It Pa /boot/sortix.initrd
system
.Xr initrd 7
.It Pa /boot/sortix.initrd.sysmerge.orig
system
.Xr initrd 7
(backup)
.It Pa /etc/machine
processor platform of this installation
.It Pa /etc/sortix-release
the current system release
.It Pa /etc/upgrade.conf
controls the bootloader upgrade behavior (see
.Xr upgrade.conf 5 )
.It Pa /sysmerge
pending upgrade is stored here
.El
.Sh SEE ALSO
.Xr development 7 ,
.Xr initrd 7 ,
.Xr installation 7 ,
.Xr kernel 7 ,
.Xr upgrade 7 ,
.Xr sysinstall 8 ,
.Xr sysupgrade 8


sysinstall/sysmerge.c
/*******************************************************************************

    Copyright(C) Jonas 'Sortie' Termansen 2016.

    This program is free software: you can redistribute it and/or modify it
    under the terms of the GNU General Public License as published by the Free
    Software Foundation, either version 3 of the License, or (at your option)
    any later version.

    This program is distributed in the hope that it will be useful, but WITHOUT
    ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
    FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
    more details.

    You should have received a copy of the GNU General Public License along with
    this program. If not, see <http://www.gnu.org/licenses/>.

    sysmerge.c
    Upgrade current operating system from a sysroot.

*******************************************************************************/

#include <sys/types.h>

#include <err.h>
#include <fcntl.h>
#include <stdbool.h>
#include <stddef.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#include "conf.h"
#include "execute.h"
#include "fileops.h"
#include "manifest.h"
#include "release.h"

static void compact_arguments(int* argc, char*** argv)
{
	for ( int i = 0; i < *argc; i++ )
	{
		while ( i < *argc && !(*argv)[i] )
		{
			for ( int n = i; n < *argc; n++ )
				(*argv)[n] = (*argv)[n+1];
			(*argc)--;
		}
	}
}

static bool has_pending_upgrade(void)
{
	return access_or_die("/boot/sortix.bin.sysmerge.orig", F_OK) == 0 ||
	       access_or_die("/boot/sortix.initrd.sysmerge.orig", F_OK) == 0 ||
	       access_or_die("/sysmerge", F_OK) == 0;
}

static void help(FILE* fp, const char* argv0)
{
	fprintf(fp, "Usage: %s [OPTION]... SOURCE\n", argv0);
	fprintf(fp, "Merge the files from SOURCE onto the current system.\n");
}

static void version(FILE* fp, const char* argv0)
{
	fprintf(fp, "%s (Sortix) %s\n", argv0, VERSIONSTR);
	fprintf(fp, "License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>.\n");
	fprintf(fp, "This is free software: you are free to change and redistribute it.\n");
	fprintf(fp, "There is NO WARRANTY, to the extent permitted by law.\n");
}

int main(int argc, char* argv[])
{
	setvbuf(stdout, NULL, _IOLBF, 0); // Pipes.

	bool cancel = false;
	bool booting = false;
	bool wait = false;

	const char* argv0 = argv[0];
	for ( int i = 1; i < argc; i++ )
	{
		const char* arg = argv[i];
		if ( arg[0] != '-' || !arg[1] )
			continue;
		argv[i] = NULL;
		if ( !strcmp(arg, "--") )
			break;
		if ( arg[1] != '-' )
		{
			char c;
			while ( (c = *++arg) ) switch ( c )
			{
			case 'w': wait = true; break;
			default:
				fprintf(stderr, "%s: unknown option -- '%c'\n", argv0, c);
				help(stderr, argv0);
				exit(1);
			}
		}
		else if ( !strcmp(arg, "--help") )
			help(stdout, argv0), exit(0);
		else if ( !strcmp(arg, "--version") )
			version(stdout, argv0), exit(0);
		else if ( !strcmp(arg, "--cancel") )
			cancel = true;
		else if ( !strcmp(arg, "--booting") )
			booting = true;
		else if ( !strcmp(arg, "--wait") )
			wait = true;
		else
		{
			fprintf(stderr, "%s: unknown option: %s\n", argv0, arg);
			help(stderr, argv0);
			exit(1);
		}
	}

	compact_arguments(&argc, &argv);

	const char* source;
	if ( cancel )
	{
		source = NULL;
		if ( 1 < argc )
			err(2, "Unexpected extra operand `%s'", argv[2]);
	}
	else if ( booting )
	{
		source = "/sysmerge";
		if ( 1 < argc )
			err(2, "Unexpected extra operand `%s'", argv[2]);
	}
	else
	{
		if ( argc < 2 )
			err(2, "No source operand was given");
		source = argv[1];
		if ( 2 < argc )
			err(2, "Unexpected extra operand `%s'", argv[2]);
	}

	if ( booting )
	{
	}
	else if ( has_pending_upgrade() )
	{
		rename("/boot/sortix.bin.sysmerge.orig", "/boot/sortix.bin");
		rename("/boot/sortix.initrd.sysmerge.orig", "/boot/sortix.initrd");
		execute((const char*[]) { "rm", "-rf", "/sysmerge", NULL }, "");
		execute((const char*[]) { "update-initrd", NULL }, "_e");
		printf("Cancelled pending system upgrade.\n");
	}
	else if ( cancel )
		printf("No system upgrade was pending.\n");

	if ( cancel )
		return 0;

	const char* old_release_path = "/etc/sortix-release";
	struct release old_release;
	if ( !os_release_load(&old_release, old_release_path, old_release_path) )
		exit(2);

	char* new_release_path;
	if ( asprintf(&new_release_path, "%s/etc/sortix-release", source) < 0 )
		err(2, "asprintf");
	struct release new_release;
	if ( !os_release_load(&new_release, new_release_path, new_release_path) )
		exit(2);
	free(new_release_path);

	// TODO: Check if /etc/machine matches the current architecture.
	// TODO: Check for version (skipping, downgrading).

	struct conf conf;
	load_upgrade_conf(&conf, "/etc/upgrade.conf");

	bool can_run_old_abi = old_release.abi_major == new_release.abi_major &&
	                       old_release.abi_minor <= new_release.abi_minor;
	if ( !can_run_old_abi && !wait )
	{
		printf("Incompatible %lu.%lu -> %lu.%lu ABI transition, "
		       "delaying upgrade to next boot.\n",
		       old_release.abi_major, old_release.abi_major,
		       new_release.abi_major, new_release.abi_major);
		wait = true;
	}

	const char* target;
	if ( wait )
	{
		printf("Scheduling upgrade to %s on next boot using %s:\n",
		       new_release.pretty_name, source);
		target = "/sysmerge";
		if ( mkdir(target, 0755) < 0 )
			err(2, "%s", target);
		execute((const char*[]) { "tix-collection", "/sysmerge", "create",
		                          NULL }, "_e");
	}
	else
	{
		printf("Upgrading to %s using %s:\n", new_release.pretty_name, source);
		target = "";
	}

	install_manifest("system", source, target);
	install_ports(source, target);

	if ( wait )
	{
		printf(" - Scheduling upgrade on next boot...\n");
		execute((const char*[]) { "cp", "/boot/sortix.bin",
		                                "/boot/sortix.bin.sysmerge.orig" }, "_e");
		execute((const char*[]) { "cp", "/boot/sortix.initrd",
		                                "/boot/sortix.initrd.sysmerge.orig" }, "_e");
		execute((const char*[]) { "cp", "/sysmerge/boot/sortix.bin",
		                                "/boot/sortix.bin" }, "_e");
		execute((const char*[]) { "/sysmerge/sbin/update-initrd", NULL }, "_e");

		printf("The system will be upgraded to %s on the next boot.\n",
		       new_release.pretty_name);
		printf("Run %s --cancel to cancel the upgrade.\n", argv[0]);

		return 0;
	}

	if ( booting )
	{
		unlink("/boot/sortix.bin.sysmerge.orig");
		unlink("/boot/sortix.initrd.sysmerge.orig");
		execute((const char*[]) { "rm", "-rf", "/sysmerge", NULL }, "");
	}

	if ( !wait && access_or_die("/etc/fstab", F_OK) == 0 )
	{
		printf(" - Creating initrd...\n");
		execute((const char*[]) { "update-initrd", NULL }, "_e");

		if ( conf.grub )
		{
			// TODO: Figure out the root device.
			//printf(" - Installing bootloader...\n");
			//execute((const char*[]) { "grub-install", "/", NULL }, "_eqQ");
			printf(" - Configuring bootloader...\n");
			execute((const char*[]) { "update-grub", NULL }, "_eqQ");
		}
		else if ( access_or_die("/etc/grub.d/10_sortix", F_OK) == 0 )
		{
			printf(" - Creating bootloader fragment...\n");
			execute((const char*[]) { "/etc/grub.d/10_sortix", NULL }, "_eq");
		}
	}

	printf("Successfully upgraded to %s.\n", new_release.pretty_name);

	return 0;
}
